File: apps/server/src/healthee/read/routine.py
"""What the owner actually did today: open fast, meditation, workouts, manual logs.

Split out of ``read/recovery.py`` (which had grown to four unrelated payloads): this
is the Today page's activity/routine roll-up, not a recovery signal, and it has its
own reason to change. All v2-native reads over ``manual_entry`` / ``workout``, on the
caller's RLS-scoped cursor.
"""

from __future__ import annotations

from datetime import UTC, date, datetime, timedelta
from uuid import UUID

from healthee.core.tenancy import reference_day, user_today
from healthee.derive._common import Cur
from healthee.read.common import sport_name


def routine_today(cur: Cur, user_id: UUID, tz: str, day: date | None = None) -> dict:
    """Open fast + the reference day's meditation, workouts, and manual-log counts.

    The three counts are per-calendar-day rows and answer for any day. ``open_fast`` does
    NOT, and is null on a past day — see :func:`_open_fast`.
    """
    as_of = reference_day(day, tz)
    return {
        "open_fast": _open_fast(cur, user_id, tz, as_of),
        "meditation_today": _meditation_today(cur, user_id, tz, as_of),
        "workouts": _workouts_today(cur, user_id, tz, as_of),
        "logs_summary": _logs_summary(cur, user_id, tz, as_of),
    }


def _open_fast(cur: Cur, user_id: UUID, tz: str, as_of: date) -> dict | None:
    """The fast running right now, or None — including on every past day.

    **Withheld rather than reconstructed, and the reason is the record, not the effort.**
    An open fast is a row with ``end_ts IS NULL`` and an ``elapsed_min`` measured against
    the clock. Neither part survives a move backwards: nothing records when the row's
    ``end_ts`` became null, so a fast closed last Tuesday is indistinguishable today from
    one that was already open on the day being asked about, and an elapsed measured from
    now under a June date is the future leak in a single field.

    An answer for a past day therefore has nothing to say here, which is what None means
    everywhere else in this payload. It is not a dash and not a zero: the app renders no
    fast, exactly as it does for an owner who is not fasting.
    """
    if as_of != user_today(tz):
        return None
    cur.execute(
        "SELECT id, ts FROM manual_entry WHERE user_id = %s AND kind='fasting' "
        "AND end_ts IS NULL ORDER BY ts DESC LIMIT 1",
        (user_id,),
    )
    row = cur.fetchone()
    if not row:
        return None
    elapsed = int((datetime.now(tz=UTC) - row[1]).total_seconds() // 60)
    return {"id": str(row[0]), "start_iso": row[1].isoformat(), "elapsed_min": elapsed}


def _meditation_today(cur: Cur, user_id: UUID, tz: str, as_of: date) -> dict:
    cur.execute(
        "SELECT COUNT(*), COALESCE(SUM(amount),0) FROM manual_entry "
        "WHERE user_id = %s AND kind='meditation' AND (ts AT TIME ZONE %s)::date = %s",
        (user_id, tz, as_of),
    )
    c, m = cur.fetchone() or (0, 0)
    return {"count": int(c or 0), "minutes": int(m or 0)}


def _workouts_today(cur: Cur, user_id: UUID, tz: str, as_of: date) -> list[dict]:
    """The reference day's device workouts (>=10 min), newest first."""
    cur.execute(
        "SELECT start_ts, duration_s, sport FROM workout WHERE user_id = %s "
        "AND (start_ts AT TIME ZONE %s)::date = %s AND COALESCE(duration_s,0) >= 600 "
        "ORDER BY start_ts DESC",
        (user_id, tz, as_of),
    )
    out = []
    for start_ts, dur_s, sport in cur.fetchall():
        dur_min = round((dur_s or 0) / 60) if dur_s else None
        end_iso = (start_ts + timedelta(seconds=int(dur_s or 0))).isoformat() if start_ts else None
        out.append(
            {
                "kind": "workout",
                "type": sport_name(sport),
                "start_iso": start_ts.isoformat() if start_ts else None,
                "end_iso": end_iso,
                "duration_min": dur_min,
                "intensity": None,
                "source": "strap",
            }
        )
    return out


def _logs_summary(cur: Cur, user_id: UUID, tz: str, as_of: date) -> dict:
    cur.execute(
        "SELECT kind, COUNT(*), COALESCE(SUM(amount),0) FROM manual_entry "
        "WHERE user_id = %s AND (ts AT TIME ZONE %s)::date = %s GROUP BY kind",
        (user_id, tz, as_of),
    )
    return {k: {"count": int(c), "total": float(t or 0)} for k, c, t in cur.fetchall()}

File: apps/mobile/lib/data/models/routine.dart
/// What the owner actually did today: workouts, meditation, and an open fast.
///
/// `read/routine.py::routine_today`, and **legacy referenced it in no file at
/// all** — the server has been computing and sending it to nobody.
///
/// Everything here is a **log of something that happened**, not a derived
/// judgement, which is why none of it is a `Reading`: there is no gate to
/// refuse, and an empty day is not a refusal. [Routine.isEmpty] is the whole honesty
/// contract for this block — a day with nothing logged renders nothing, never a
/// heading over an empty list.
///
/// ## `logs_summary`, and the day that rendered as empty while the wire said otherwise
///
/// `read/routine.py` has always shipped `logs_summary` — the per-kind `{count, total}`
/// roll-up of the day's manual entries — and `grep logs_summary apps/mobile/lib` returned
/// nothing. That was not merely dead weight: [Routine.isEmpty] was
/// `workouts.isEmpty && meditationCount == 0 && openFast == null`, so a day whose only
/// entry was caffeine, or hydration, or any kind but meditation drew **no journal panel
/// at all** while the payload was explicitly reporting that entry.
///
/// Nothing was fabricated — the panel was silent, not wrong — but it is the payload
/// saying less than it knows, and a key nothing reads invites the next reader to assume
/// it is drawn. [Routine.logs] now follows the wire, and [Routine.isEmpty] with it.
///
/// The kinds already drawn from their own fields are **not** drawn twice:
/// `logs_summary` counts every `manual_entry` kind including `meditation` and `fasting`,
/// which have dedicated blocks above. [Routine.otherLogs] is the rest.
library;

import 'package:meta/meta.dart';

/// One session the strap or a manual log recorded today.
@immutable
class RoutineEvent {
  /// A logged session.
  const RoutineEvent({
    required this.kind,
    required this.type,
    required this.startIso,
    required this.durationMin,
    required this.source,
  });

  /// Parses one entry of `routine.workouts`.
  factory RoutineEvent.fromJson(Map<String, Object?> json) {
    return RoutineEvent(
      kind: json['kind'] as String? ?? 'workout',
      type: json['type'] as String?,
      startIso: json['start_iso'] as String?,
      durationMin: (json['duration_min'] as num?)?.toInt(),
      source: json['source'] as String?,
    );
  }

  /// `workout`.
  final String kind;

  /// The sport, as the server named it — `Outdoor run`.
  final String? type;

  /// When it started, ISO-8601.
  final String? startIso;

  /// How long it ran. The server drops anything under ten minutes.
  final int? durationMin;

  /// Where it came from — `strap`, or a manual log. Named on the wire because
  /// the two are different instruments and a session's provenance is part of it.
  final String? source;
}

/// A fast that has begun and not yet ended.
@immutable
class OpenFast {
  /// An in-progress fast.
  const OpenFast({required this.startIso, required this.elapsedMin});

  /// Parses `routine.open_fast`, or null when none is running.
  static OpenFast? maybe(Object? raw) {
    if (raw is! Map<String, Object?>) {
      return null;
    }
    final elapsed = (raw['elapsed_min'] as num?)?.toInt();
    if (elapsed == null) {
      return null;
    }
    return OpenFast(startIso: raw['start_iso'] as String?, elapsedMin: elapsed);
  }

  /// When it began, ISO-8601.
  final String? startIso;

  /// How long it has been running, in minutes. The server computes this against
  /// its own clock, which is why the app does not recompute it from [startIso]
  /// — two answers to "how long" is one of them being wrong.
  final int elapsedMin;
}

/// One kind of manual entry, rolled up for the day: how many, and how much.
@immutable
class LogTally {
  /// A day's total for one kind.
  const LogTally({required this.kind, required this.count, required this.total});

  /// The entry kind as the server names it — `caffeine`, `water`, `mood`, …
  final String kind;

  /// How many entries of this kind were logged.
  final int count;

  /// Their summed `amount`. Zero for a kind that carries no amount (`mood`,
  /// `symptom`, `habit`), which is why [count] is the figure that decides whether
  /// there is anything to say.
  final double total;
}

/// Today's routine: sessions, meditation, an open fast, and the day's other logs.
@immutable
class Routine {
  /// Builds a day. Prefer [Routine.fromJson].
  const Routine({
    required this.workouts,
    required this.meditationCount,
    required this.meditationMinutes,
    required this.openFast,
    this.logs = const <LogTally>[],
  });

  /// Parses the `routine` block. Never null — an absent block is an empty day,
  /// and [isEmpty] is what decides whether anything is drawn.
  factory Routine.fromJson(Map<String, Object?> json) {
    final meditation = json['meditation_today'] is Map<String, Object?>
        ? json['meditation_today']! as Map<String, Object?>
        : const <String, Object?>{};
    return Routine(
      workouts: [
        for (final entry in (json['workouts'] as List? ?? const []))
          if (entry is Map<String, Object?>) RoutineEvent.fromJson(entry),
      ],
      meditationCount: (meditation['count'] as num?)?.toInt() ?? 0,
      meditationMinutes: (meditation['minutes'] as num?)?.toInt() ?? 0,
      openFast: OpenFast.maybe(json['open_fast']),
      logs: _tallies(json['logs_summary']),
    );
  }

  /// Parses `logs_summary` — `{kind: {count, total}}` — in the server's own key order.
  ///
  /// A kind with a zero or missing count is dropped rather than kept at zero: the roll-up
  /// is a GROUP BY, so a zero cannot arrive from the server, and carrying one would put a
  /// row on screen for something the owner did not log.
  static List<LogTally> _tallies(Object? raw) {
    if (raw is! Map<String, Object?>) {
      return const <LogTally>[];
    }
    final out = <LogTally>[];
    for (final entry in raw.entries) {
      final value = entry.value;
      if (value is! Map<String, Object?>) {
        continue;
      }
      final count = (value['count'] as num?)?.toInt() ?? 0;
      if (count <= 0) {
        continue;
      }
      out.add(
        LogTally(
          kind: entry.key,
          count: count,
          total: (value['total'] as num?)?.toDouble() ?? 0,
        ),
      );
    }
    return out;
  }

  /// Today's sessions, newest first.
  final List<RoutineEvent> workouts;

  /// How many meditation entries were logged today.
  final int meditationCount;

  /// Their total minutes.
  final int meditationMinutes;

  /// A fast still running, or null.
  final OpenFast? openFast;

  /// Every manual-entry kind the day carried, rolled up. `routine.logs_summary`.
  final List<LogTally> logs;

  /// The logged kinds that do NOT already have a block of their own above.
  ///
  /// `logs_summary` counts every `manual_entry` kind, so `meditation` is in it and so is
  /// `fasting` — both already drawn from their own fields. Printing them from here as
  /// well would be two renderings of one entry, disagreeing the moment the two shapes
  /// diverge (the fast's block is about an OPEN one; the tally counts every fast of the
  /// day). One definition per thing on screen.
  Iterable<LogTally> get otherLogs =>
      logs.where((tally) => !_ownBlock.contains(tally.kind));

  static const Set<String> _ownBlock = <String>{'meditation', 'fasting'};

  /// Whether there is anything at all to say. **A day with nothing logged draws
  /// nothing** — no card, no heading, no "nothing yet" placeholder.
  ///
  /// It follows the WIRE: a day whose only entry was caffeine used to render no panel
  /// while the payload reported that entry (audit C6).
  bool get isEmpty =>
      workouts.isEmpty &&
      meditationCount == 0 &&
      openFast == null &&
      otherLogs.isEmpty;
}
